x-pack/plugins/fleet/server/services/actions/actions.ts

```diff
@@ -0,0 +1,252 @@
+    throw new FleetActionsError(
+      `Error creating action: ${createActionError.message}`,
+      createActionError
+    );
+  }
+};
+
+type BulkCreate = Array<{ create: { _index: string; _id: string } } | object>;
+
+export const bulkCreateActions = async (
+  esClient: ElasticsearchClient,
+  _actions: Array<Partial<FleetAction>>
+): Promise<
+  Array<{
+    id: string;
+    action: Partial<FleetAction>;
+  }>
+> => {
+  const actions: Array<Partial<FleetAction>> = [];
+  const bulkCreateActionsBody = _actions.reduce<BulkCreate>((acc, action) => {
+    acc.push({ create: { _index: AGENT_ACTIONS_INDEX, _id: action.action_id ?? uuidV4() } });
+    const actionDoc = {
+      ...action,
+      action_id: action.action_id ?? uuidV4(),
+      '@timestamp': action['@timestamp'] ?? new Date().toISOString(),
+    };
+    acc.push(actionDoc);
+    actions.push(actionDoc);
+
+    return acc;
+  }, []);
+
+  try {
+    const bulkCreateActionsResponse = (await esClient.bulk(
+      {
+        body: bulkCreateActionsBody,
+        refresh: 'wait_for',
+      },
+      {
+        meta: true,
+      }
+    )) as unknown as estypes.BulkResponse;
+
+    const [errorItems, successItems] = partition(
+      bulkCreateActionsResponse.items,
+      (a) => a.create?.error
+    );
+
+    successItems.forEach((successItem) => {
+      auditLoggingService.writeCustomAuditLog({
+        message: `User created Fleet action [id=${successItem.create?._id}]`,
+      });
+    });
+
+    // log errored items
+    if (errorItems.length) {
+      appContextService.getLogger().error(new FleetActionsError(JSON.stringify(errorItems)));
+    }
+
+    // return success items
+    return successItems.map((item, i) => ({
+      id: item.create!._id as string,
+      action: actions[i],
+    }));
+  } catch (createBulkActionsError) {
+    throw new FleetActionsError(
+      `Error creating bulk actions: ${createBulkActionsError.message}`,
+      createBulkActionsError
+    );
+  }
+};
+
+export const getActionsByIds = async (
+  esClient: ElasticsearchClient,
+  actionIds: string[]
+): Promise<FleetAction[]> => {
+  try {
+    const getActionsResponse = await esClient.search({
+      index: AGENT_ACTIONS_INDEX,
+      query: {
+        bool: {
+          filter: [
+            {
+              terms: {
+                action_id: actionIds,
+              },
+            },
+          ],
+        },
+      },
+    });
+    return getActionsResponse.hits?.hits?.map((hit) => hit._source ?? {}) as FleetAction[];
+  } catch (getActionsByIdError) {
+    throw new FleetActionsError(
+      `Error getting action: ${getActionsByIdError.message}`,
+      getActionsByIdError
+    );
+  }
+};
+
+export const getActionsWithKuery = async (
+  esClient: ElasticsearchClient,
+  kuery: string
+): Promise<{ actions: FleetAction[]; total: number }> => {
+  try {
+    const query = toElasticsearchQuery(fromKueryExpression(kuery));
+    const getActionSearchResponse = await esClient.search({
```

Comment:
Without pagination this will only return (I think) the first 10 items found (I think that is the default for `search()`)

Reply:
6484cd1963a51303638fefdc701c5f249f6f58f9